src/Coinbase/Exchange/Types/Socket.hs

```diff
@@ -122,225 +401,323 @@ data ExchangeMessage
 instance ToJSON ExchangeMessage where
-    toJSON Open{..} = object
-        [ "type"       .= ("open" :: Text)
-        , "time"       .= msgTime
-        , "product_id" .= msgProductId
-        , "sequence"   .= msgSequence
-        , "order_id"   .= msgOrderId
-        , "side"       .= msgSide
-        , "remaining_size" .= msgRemainingSize
-        , "price"      .= msgPrice
-        ]
-    toJSON Done{..} = object
-        ([ "type"      .= ("done" :: Text)
-        , "time"       .= msgTime
-        , "product_id" .= msgProductId
-        , "sequence"   .= msgSequence
-        , "order_id"   .= msgOrderId
-        , "side"       .= msgSide
-        , "reason"     .= msgReason
-        ]
-        ++ case msgMaybePrice of
-                Nothing -> []
-                Just  p -> ["price" .= p]
+  toJSON ErrorMsg{..} = object
+    [ "type" .= ("error" :: Text)
+    , "message" .= msgErrorMessage
+    ]
+  toJSON SubscriptionsMsg{..} = object
+    [ "type" .= ("subscriptions" :: Text)
+    ,  "channels" .=  msgSubscriptionsMsgs
+    ]
+  toJSON HeartbeatMsg{..} = object
+    [ "type"          .= ("heartbeat" :: Text)
+    , "sequence"      .= msgSequence
+    , "time"          .= msgTime
+    , "product_id"    .= msgProductId
+    , "last_trade_id" .= msgLastTradeId
+    ]
+  toJSON StatusMsg{..} = object
+    [ "type"       .= ("status" :: Text)
+    , "products"   .= msgProducts
+    , "currencies" .= msgCurrencies
+    ]
+  toJSON StartTickerMsg{..} = object
+    [ "type"       .= ("ticker" :: Text)
+    , "sequence"   .= msgSequence
+    , "product_id" .= msgProductId
+    , "price"      .= msgPrice
+    , "best_ask"   .= msgBestAsk
+    , "best_bid"   .= msgBestBid
+    ]
+  toJSON TickerMsg{..} = object
+    ( [ "type"       .= ("ticker" :: Text)
+      , "time"       .= msgTime
+      , "sequence"   .= msgSequence
+      , "product_id" .= msgProductId
+      , "trade_id"   .= msgTradeId
+      , "price"      .= msgPrice
+      , "side"       .= msgSide
+      , "last_size"  .= msgLastSize
+      ] ++ case msgBestAsk of
+            Nothing      -> []
+            Just bestAsk -> ["best_ask" .= bestAsk]
+        ++ case msgBestBid of
+            Nothing      -> []
+            Just bestBid -> ["best_bid" .= bestBid]
+    )
+  toJSON L2SnapshotMsg{..} = object
+    [ "type"       .= ("snapshot" :: Text)
+    , "product_id" .= msgProductId
+    , "asks"       .= msgAsks
+    , "bids"       .= msgBids
+    ]
+  toJSON L2UpdateMsg{..} = object
+    [ "type"       .= ("l2update" :: Text)
+    , "time"       .= msgTime
+    , "product_id" .= msgProductId
+    , "changes"    .= msgChanges
+    ]
+  toJSON OpenMsg{..} = object
+    [ "type"       .= ("open" :: Text)
+    , "sequence"   .= msgSequence
+    , "time"       .= msgTime
+    , "product_id" .= msgProductId
+    , "order_id"   .= msgOrderId
+    , "side"       .= msgSide
+    , "remaining_size" .= msgRemainingSize
+    , "price"      .= msgPrice
+    ]
```

Comment:
I would like to be able to use `deriving via` to derive these automatically. This is technical debt. And it seems so close now! (Just change from camelCase to snake_case and remove the prefix) Some future work.

Reply:
I've never used `deriving via`, though I know the basic idea. How does that work here?

Reply:
I am not sure exactly. We would have to write a rule (i.e. a toJSON instance for a newtype wrapper) to derive these and then just apply this one rule to all these types. But I am not sure how to deal with the `"type"` field in this context. It seems worth exploring now. 

The best description I know for deriving via is here:
https://skillsmatter.com/skillscasts/12654-deriving-via

Reply:
It seems like this should be straightforward using Aeson's generic-deriving facilities and applying an appropriate [`fieldLabelModifier`](http://hackage.haskell.org/package/aeson-1.0.1.0/docs/Data-Aeson-Types.html#v:fieldLabelModifier) and [`constructorTagModifier`](http://hackage.haskell.org/package/aeson-1.0.1.0/docs/Data-Aeson-Types.html#v:constructorTagModifier). However, I'm pretty sure that I would have done that if possible. Perhaps there is some problem in deriving generic representations?

I'll look at it.

Reply:
OK, I've looked at this and I think it can indeed be done with `genericToJSON` and a minimum of fiddling. I will re-write and resubmit tomorrow.

Reply:
I take it back. This can *almost* be done by deriving generically, needing only to change a bunch of field and constructor tag names, but there are a small handful of cases where the fields on the `ExchangeMessage` constructors are insufficiently parallel to the JSON messages to make this work.

[I've included a few footnotes below on some digressions from the main point.]

For example, the fact that "received" messages are parsed into two constructors (`ReceivedLimitMsg` and `ReceivedMarketMsg`) makes it impossible to derive generically because there's no way to add JSON fields that don't correspond to a Haskell record field (the "order_type" field is absent in generically produces JSON because that information is captured by the distinction between the two constructors themselves).[1]

Another example of insufficiently parallel types and messages: "received" messages for market orders sometimes have a "size" field and sometimes a "funds" fields, but the `ReceivedMarketMsg` constructor has a combined `msgMarketBounds` field with type `Either Size (Maybe Size, Cost)`. This is explained by a comment in the code which reads:

> market orders have no price and are bounded by either size, funds or both

Evidently this is meant to capture the idea that a market order is specified either with a quantity (`Size`) to sell or an amount of funds (`Cost`) to use to buy (evidently with the possibility in the latter case of there also sometimes being a quantity present).[2]

So we can either (a) leave things as they are, with a big hand-coded `ToJSON` instance; or (b) re-write the data structure to be parallel to the server's JSON and then derive generically. In the above example, that would mean combining the `ReceivedLimitMsg` and `ReceivedMarketMsg` constructors and replacing the `msgMarketBounds :: Either Size (Maybe Size, Cost)` field with two fields, `msgMaybePrice :: Maybe Price` and `msgMaybeFunds :: Maybe Cost`.[3]

In my opinion, it makes sense to re-write the data structure to be parallel to the JSON (at least in the long run, if not now, and at least in the case of `ExchangeMessage`, though maybe not of other types). That would vastly simplify the code and eliminate pretty much all of the pain of debugging the Aeson instances. And I think it makes sense: the point of this library is to implement the Coinbase Pro API, which means giving users a good Haskell representation of what Coinbase's servers spit out. To me, the fact that we have to write fiddly `ToJSON` and `FromJSON` instances by hand suggests that our types are trying to do too much.

Now, I can see the value in providing types that express stronger invariants than what the API makes explicit in its message schema, but it doesn't seem to me that the existing types, where they diverge from the schema in ways like I mentioned above, actually help much. Is that `Either` type for a market order's "bounds" (a concept not present in the API or its documentation) really better than just using a `Maybe` for the two message fields that don't always show up? Even if the `Either` correctly encodes the permutations of fields that you might see, who cares? It mostly just seems more complicated.

There's something of a middle ground, which I imagine is similar to what you had in mind when you mentioned `deriving via`: we can use exactly parallel types for parsing, then translate into non-parallel types like those we have, e.g., splitting the representation of the "received" message type into two constructors. To the extent that we feel non-parallel representations really add value that might make sense.

Thoughts?

-----
Footnotes:

1. I mean it's impossible to do this generically with the provided facilities for modifications along the way, which only allow for re-naming things. Obviously we could produce some wrong or incomplete JSON generically, then go back and fix it by adding fields or whatever. I mention this possibility towards the end of this comment.
2. Is this another Coinbase bug, or is this just wrong? A market buy should always specify an amount of funds (aka "cost"), without specifying a quantity (aka "size"), and a market sell should always specify a quantity ("size") without specifying an amount of funds ("cost"). Do you know the origin of this code? If not, this may merit further investigation.
3. Why `Cost` and not`Funds`? This reminds me that `Size`, `Cost`, and other types that refer to quantities of some asset would benefit from having the asset (aka "product") name or ID embedded in the type.

Reply:
Whoa! There's a lot to unpack on your comments. Here are my thoughts.

First and foremost, I am *not* as concerned about deriving the `ToJSON` instances as I am about deriving the `Arbitrary` instances. This is why I said this is maybe future work. I think it is worth improving, but I don't think it (the technical debt) has to be solved now, for these `ToJSON` instances.

You did a good job at pointing out and documenting the difficulties with deriving them automatically because of the multiple constructors (i.e. `ReceivedLimitMsg` and `ReceivedMarketMsg`).

I strongly disagree that there is no value in capturing the API invariants expressed in the type of the field `msgMarketBounds :: Either Size (Maybe Size, Cost)`. This invariant does exist (it just can't be expressed in JSON), This is evidenced by the fact that I don't ever recall a parsing failure because of these (let me know if you see one). Furthermore, once this invariant is established, we don't have to keep checking this value again and again as would be the case with multiple `Maybe`s. I think this is one of those ["Parse, don't validate"](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/) and "make invalid states unrepresentable" situations. Does this type make it more complicated to parse? For sure! But precisely because it enforces more strict invariants. (I have written other code that depends on this library and it was quite useful to be able to count on these invariants. The alternative is having to do this "type checking" *again* at a later stage. The invariant does show up either way.)

Although it may seem initially counter intuitive, allowing market orders to be bound by size, funds or both, is not a bug. This is exactly how market orders should be implemented in a "continuous limit orderbook" market. It is implicit that when limited by both, we should use the minimum (most restrictive) bound. This is one situation where Coinbase got it right and almost everybody else got it wrong.

The type `Cost` was created to distinguish from `Price`. The `Cost` of 2 (BTC) at a price of 10K (USD/BTC) is 20K (USD). Note the units. `Funds` seems to also match this idea, but seems a little bit less precise. I agree that costs could have the units encoded at the type level, for example, as a phantom type. As a requirement, I think that any implementation of this should also take care of the above calculation and changing units from Price to Cost.

Considering that after you looked at it carefully, it looks like a significant amount of work for an uncertain amount of gain. I suggest we leave it as is for now.